src/webgpu/shader/execution/shader_io/compute_builtins.spec.ts

```diff
@@ -254,3 +255,505 @@ g.test('inputs')
+fn main(@builtin(subgroup_size) size : u32,
+        @builtin(workgroup_id) wgid : vec3u,
+        @builtin(local_invocation_index) lid : u32) {
+  output[lid + wgid.x * stride] = size;
+  let ballot = countOneBits(subgroupBallot(true));
+  let ballotSize = ballot[0] + ballot[1] + ballot[2] + ballot[3];
+  compare[lid + wgid.x * stride] = ballotSize;
+}`;
+
+    const numInvocations = wgThreads * t.params.numWGs;
+    const sizesBuffer = t.makeBufferWithContents(
+      new Uint32Array([...iterRange(numInvocations, x => 0)]),
+      GPUBufferUsage.STORAGE | GPUBufferUsage.COPY_SRC | GPUBufferUsage.COPY_DST
+    );
+    t.trackForCleanup(sizesBuffer);
+    const compareBuffer = t.makeBufferWithContents(
+      new Uint32Array([...iterRange(numInvocations, x => 0)]),
+      GPUBufferUsage.STORAGE | GPUBufferUsage.COPY_SRC | GPUBufferUsage.COPY_DST
+    );
+    t.trackForCleanup(compareBuffer);
+
+    const pipeline = t.device.createComputePipeline({
+      layout: 'auto',
+      compute: {
+        module: t.device.createShaderModule({
+          code: wgsl,
+        }),
+        entryPoint: 'main',
+      },
+    });
+    const bg = t.device.createBindGroup({
+      layout: pipeline.getBindGroupLayout(0),
+      entries: [
+        {
+          binding: 0,
+          resource: {
+            buffer: sizesBuffer,
+          },
+        },
+        {
+          binding: 1,
+          resource: {
+            buffer: compareBuffer,
+          },
+        },
+      ],
+    });
+
+    const encoder = t.device.createCommandEncoder();
+    const pass = encoder.beginComputePass();
+    pass.setPipeline(pipeline);
+    pass.setBindGroup(0, bg);
+    pass.dispatchWorkgroups(t.params.numWGs, 1, 1);
+    pass.end();
+    t.queue.submit([encoder.finish()]);
+
+    const sizesReadback = await t.readGPUBufferRangeTyped(sizesBuffer, {
+      srcByteOffset: 0,
+      type: Uint32Array,
+      typedLength: numInvocations,
+      method: 'copy',
+    });
+    const sizesData: Uint32Array = sizesReadback.data;
+
+    const compareReadback = await t.readGPUBufferRangeTyped(compareBuffer, {
+      srcByteOffset: 0,
+      type: Uint32Array,
+      typedLength: numInvocations,
+      method: 'copy',
+    });
+    const compareData: Uint32Array = compareReadback.data;
+
+    t.expectOK(checkSubgroupSizeConsistency(sizesData, compareData, minSize, maxSize, wgThreads));
+  });
+
+/**
+ * Checks the consistency of subgroup_invocation_id builtin values
+ *
+ * @param data The subgroup_invocation_id output data
+ * @param
+ */
+function checkSubgroupInvocationIdConsistency(
+  data: Uint32Array,
+  ids: Uint32Array,
+  subgroupSize: number,
+  invocations: number,
+  numWGs: number
+): Error | undefined {
+  for (let wg = 0; wg < numWGs; wg++) {
+    const mappings = new Map();
+    for (let i = 0; i < invocations; i++) {
+      const idx = i + invocations * wg;
+      const subgroup_id = ids[idx];
+      if (subgroup_id === 999) {
```

Comment:
I would recommend making magic number `999` here and line 675 to a named constant, e.g.
``` typescript
// Initializer for array<u32> buffer holding subgroup id mapped within the shader,
// the value is impossible to be outputted by the shader.
const subgroupIdBufferInitializer: number = 0xffffffff;
```